src/System.Security.Cryptography.Algorithms/src/System/Security/Cryptography/HKDF.cs

```diff
@@ -0,0 +1,255 @@
+
+            int hashLength = HashLength(hashAlgorithmName);
+
+            if (outputLength <= 0 || outputLength > 255 * hashLength)
+                throw new ArgumentOutOfRangeException(nameof(outputLength));
+
+            info = info ?? Array.Empty<byte>();
+            byte[] result = new byte[outputLength];
+            Expand(hashAlgorithmName, hashLength, prk, result, info);
+
+            return result;
+        }
+
+        /// <summary>
+        /// Performs the HKDF-Expand function
+        /// See section 2.3 of <a href="https://tools.ietf.org/html/rfc5869#section-2.3">RFC5869</a>
+        /// </summary>
+        /// <param name="hashAlgorithmName">Hash algorithm used for HMAC operations</param>
+        /// <param name="prk">Pseudorandom key of at least <see cref="HashLength"/> bytes (usually the output from Expand step)</param>
+        /// <param name="output">Output buffer representing output keying material</param>
+        /// <param name="info">Context and application specific information (can be an empty span)</param>
+        public static void Expand(HashAlgorithmName hashAlgorithmName, ReadOnlySpan<byte> prk, Span<byte> output, ReadOnlySpan<byte> info)
+        {
+            int hashLength = HashLength(hashAlgorithmName);
+
+            // We want to throw different exception type for this overload
+            if (output.Length > 255 * hashLength)
+                throw new ArgumentException(nameof(output));
+
+            Expand(hashAlgorithmName, hashLength, prk, output, info);
+        }
+
+        private static void Expand(HashAlgorithmName hashAlgorithmName, int hashLength, ReadOnlySpan<byte> prk, Span<byte> output, ReadOnlySpan<byte> info)
+        {
+            if (prk.Length < hashLength)
+                throw new ArgumentException(nameof(prk));
+
+            if (output.Length > 255 * hashLength)
+                throw new ArgumentOutOfRangeException(nameof(output));
+
+            Span<byte> counter = stackalloc byte[1];
+            Span<byte> t = Span<byte>.Empty;
+            Span<byte> remainingOutput = output;
+
+            using (IncrementalHash hmac = IncrementalHash.CreateHMAC(hashAlgorithmName, prk))
+            {
+                for (int i = 1; ; i++)
+                {
+                    hmac.AppendData(t);
+                    hmac.AppendData(info);
+                    counter[0] = (byte)i;
+                    hmac.AppendData(counter);
+
+                    if (remainingOutput.Length >= hashLength)
+                    {
+                        t = remainingOutput.Slice(0, hashLength);
+                        remainingOutput = remainingOutput.Slice(hashLength);
+                        GetHashAndReset(hmac, t);
+                    }
+                    else
+                    {
+                        if (remainingOutput.Length > 0)
+                        {
+                            Span<byte> lastChunk = stackalloc byte[hashLength];
```

Comment:
Let's assert here so if we ever add a bigger hash we consider if stackalloc is still the right thing to do.

```suggestion
                            // SHA-512 is the largest supported value
                            Debug.Assert(hashLength <= 512 / 8);
                            Span<byte> lastChunk = stackalloc byte[hashLength];
```